### DISCUSSION TITLE
How do I reserve an IP address with in a subnet

### DISCUSSION THREAD

**Discussion by tskgit** [Azure/ipam#208] [General]
I would like to reserve list of  IP address with in a subnet and is there any way to view all assigned IP addresses in a subnet?

**Comment by DCMattyG** [comment]
Hey there @tskgit, thanks so much for reaching out, and welcome to Azure IPAM discussions!

At this time, there isn't a mechanism to reserve IP addresses within a Subnet, though I am always open to adding in functionality that makes sense and is useful for everyone. Can you help me better understand your use case for reserving a list of static IP's within a subnet? What is an example workflow that you'd use this step for?

To answer your second question, yes you can view all IP addresses within a Subnet today. You can do this by navigating to Discover > Endpoints, and then filtering by the vNET/Subnet name at the top of the screen like so...

![image](https://github.com/Azure/ipam/assets/12955732/1334f5a0-f421-4a2c-9518-efeb280efc5f)

Hope that helps, and looking forward to hearing more about how you plan to utilize static IP reservations within a subnet!

**Reply by tskgit** [reply]
Hi @DCMattyG,

Reserving a list of static IP addresses within a subnet can be useful in various scenarios where we need to ensure a specific IP address is always assigned to a particular device or service.

Here's an example workflow

1.Determine the specific devices or services that require reserved/static IP addresses within a subnet. For example, you might have a web server, database server, and a load balancer that need consistent IP assignments.
2.Determine the specific IP addresses that need to be reserved within the subnet. For example, you might reserve 192.168.0.10 for the web server, 192.168.0.20 for the database server, and 192.168.0.30 for the load balancer.

And is I see an option "Reservations" in the configure, I would like to know the use case of this option? 

![image](https://github.com/Azure/ipam/assets/70080893/a56f34ba-a962-4997-ae1a-e71335464413)


Also would like to export all assigned IPs in the endpoints, is there any way to export these results?

Thank you!

**Reply by DCMattyG** [reply]
Thanks for the details @tskgit!

The intent of the **Reservations** section is for carving out new Virtual Networks in Azure. One of the biggest challenges is when a Network team assigns a large block of IP space to a team in Azure (for example 10.0.0.0/16) and then the Azure team is responsible for creating **non-overlapping** Virtual Networks from that space. The Azure IPAM tool enables a self-service type interaction where users can create **Blocks** and assign the large CIDR block, and then via API new Virtual Network CIDR's can be "reserved" and then assigned accordingly within Azure. Currently, this is accomplished via API and we have the process well documented [here](https://azure.github.io/ipam/#/api/README).

As far as assigning static IP's to endpoints within a subnet, I'm curious why you would be relying on IP addresses and not FQDN's (DNS entries) which would make life far easier because the underlying IP can be "whatever" but the DNS name will remain stable. I agree that "back in the day" I've done similar myself, but in a cloud world I've learned its better to embrace the dynamic nature of things rather that try to force them to fit into previous on-premises patterns (my $0.02).

FYI: In Azure, even though IP's are assigned "dynamically", they never change once they are assigned to a vNIC...

Source: https://learn.microsoft.com/en-us/azure/virtual-network/virtual-networks-faq#what-are-the-types-of-ip-addresses-that-i-can-assign-to-vms

> "A private IP address assigned through the dynamic or static method remains assigned to a virtual machine (Resource Manager) until the resource is deleted."

I hope all of this has helped to clarify a few things for you, but as always I'm more than happy to be of service however I can. Please let me know if there's anything else I can be of assistance with as you continue exploring the Azure IPAM service.

**Reply by DCMattyG** [reply]
To answer your last question about exporting the data, currently there is no "export" function within the product (though I am actively working on a solution for this).

That said, you could easily accomplish this by calling the **Endpoints** API, filtering by vNET/Subnet, then exporting the data to CSV as shown in [this](https://www.tutorialspoint.com/how-to-convert-json-file-to-csv-file-using-powershell) example. We also have some base PowerShell examples in our [docs](https://azure.github.io/ipam/#/api/README?id=sample-api-calls) page which show how to obtain a token and call the Azure IPAM API's as well.

I'm happy to work up some more specific example code for you if needed, just let me know!

**Reply by fmorrison42** [reply]
Another example that I am currently faced with and would find Static IP Reservations useful is in VM Failovers. Here is my current situation: 
I have two web servers in a load balancer. When they failover into a different Azure Region, they are placed into a new load balancer in the new Region. However, because of the nature of web servers and load balancers, having static IP addresses just makes things run and connect.  I have tried to set static IP addresses in the Failover Settings, however, I find that randomly, that "reservation" will have randomly been taken by another resource later on and is no longer available, because the IP address wasn't ACTIVELY being used in the Failover Region.  I just need that Static IP address to sit empty if and when the day comes that these web servers have to failover into the Failover Region, rather than my team having to scramble to find an available IP address to statically assign to the VMs then update all of the rest of the scripts and documentation that keep the Load Balancer and Web Servers running on both ends.
